### PR TITLE
Add User-Agent header to OTLP exporter requests

### PR DESCRIPTION
**Pull request by jack-berg**
Implements spec [PR #2684](https://github.com/open-telemetry/opentelemetry-specification/pull/2684).

### CODE REVIEW

File: exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/ManagedChannelTelemetryExporterBuilder.java [hunk 1]
```diff
@@ -51,6 +52,10 @@ public TelemetryExporterBuilder<T> setEndpoint(String endpoint) {
     if (!uri.getScheme().equals("https")) {
       channelBuilder.usePlaintext();
     }
+    // User-Agent can only be set at the channel level with upstream gRPC client. If a user wants
+    // the User-Agent to be spec compliant they must add manually set the user agent when building
+    // their channel.
```
Comment (trask) [inline exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/ManagedChannelTelemetryExporterBuilder.java:57]:
```suggestion
    // User-Agent can only be set at the channel level with upstream gRPC client. If a user wants
    // the User-Agent to be spec compliant they must manually set the user agent when building
    // their channel.
```

File: exporters/otlp/common/src/test/java/io/opentelemetry/exporter/internal/otlp/OtlpUserAgentTest.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.concurrent.atomic.AtomicReference;
+import org.junit.jupiter.api.Test;
+
+class OtlpUserAgentTest {
+
+  @Test
+  void getUserAgent() {
+    assertThat(OtlpUserAgent.getUserAgent()).matches("OTel OTLP Exporter Java/1.*");
```
Comment (trask) [inline exporters/otlp/common/src/test/java/io/opentelemetry/exporter/internal/otlp/OtlpUserAgentTest.java:17]:
```suggestion
    assertThat(OtlpUserAgent.getUserAgent()).matches("OTel OTLP Exporter Java/1\\..*");
```

File: exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/otlp/OtlpUserAgent.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  static {
+    userAgent = "OTel OTLP Exporter Java/" + readVersion();
+  }
+
+  private static String readVersion() {
+    Properties properties = new Properties();
+    try {
+      properties.load(
+          Resource.class.getResourceAsStream(
+              "/io/opentelemetry/exporter/internal/otlp/version.properties"));
```
Comment (trask) [inline exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/otlp/OtlpUserAgent.java:29]: I almost never use relative paths for loading resources, but maybe this is why that functionality exists?
```suggestion
      properties.load(Resource.class.getResourceAsStream("version.properties"));
```

File: exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/otlp/OtlpUserAgent.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+ * This class is internal and is hence not for public use. Its APIs are unstable and can change at
+ * any time.
+ */
+public class OtlpUserAgent {
+
+  private static final String userAgent;
+
+  static {
+    userAgent = "OTel OTLP Exporter Java/" + readVersion();
+  }
```
Comment (trask) [inline exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/otlp/OtlpUserAgent.java:22]:
```suggestion
  private static final String userAgent = "OTel OTLP Exporter Java/" + readVersion();
```

File: exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/AbstractGrpcTelemetryExporterTest.java [hunk 5]
```diff
@@ -216,6 +216,14 @@ void export() {
     assertThat(exporter.export(telemetry).join(10, TimeUnit.SECONDS).isSuccess()).isTrue();
     List<U> expectedResourceTelemetry = toProto(telemetry);
     assertThat(exportedResourceTelemetry).containsExactlyElementsOf(expectedResourceTelemetry);
+
+    // Assert request contains OTLP spec compliant User-Agent header
+    assertThat(httpRequests)
+        .singleElement()
+        .satisfies(
+            req -> {
+              assertThat(req.headers().get("User-Agent")).matches("OTel OTLP Exporter Java/1.*");
```
Comment (trask) [inline exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/AbstractGrpcTelemetryExporterTest.java:225]:
```suggestion
              assertThat(req.headers().get("User-Agent")).matches("OTel OTLP Exporter Java/1\\..*");
```

File: exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/otlp/OtlpUserAgent.java (hunk 3 above)
Comment (jack-berg) [reply exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/otlp/OtlpUserAgent.java:29]: Yeah I didn't know that would work. Feels mysterious that this works, but it does and is more readable 👍

Comment (jack-berg) [reply exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/otlp/OtlpUserAgent.java:29]: Yup! I caught that and have it committed locally, just forgot to push it up (hence all the resolved comments without a corresponding commit 😬).

**Review by trask — APPROVED** [review]

**Review by mateuszrzeszutek — APPROVED** [review]

**Comment by jack-berg** [comment]
@jkwatson please take a look when you have a chnace

File: exporters/otlp/common/build.gradle.kts [hunk 6]
```diff
@@ -9,7 +9,7 @@ plugins {
 }
 
 description = "OpenTelemetry Protocol Exporter"
-otelJava.moduleName.set("io.opentelemetry.exporter.otlp.internal")
+otelJava.moduleName.set("io.opentelemetry.exporter.internal.otlp")
```
Comment (jkwatson) [inline exporters/otlp/common/build.gradle.kts:12]: what affect does this have?

File: exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/ManagedChannelTelemetryExporterBuilder.java [hunk 7]
```diff
@@ -51,6 +52,10 @@ public TelemetryExporterBuilder<T> setEndpoint(String endpoint) {
     if (!uri.getScheme().equals("https")) {
       channelBuilder.usePlaintext();
     }
+    // User-Agent can only be set at the channel level with upstream gRPC client. If a user wants
+    // the User-Agent to be spec compliant they must manually set the user agent when building
+    // their channel.
```
Comment (jkwatson) [inline exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/ManagedChannelTelemetryExporterBuilder.java:57]: Is this something that should be in docs somewhere? Or javadoc that's more visible than an internal comment?

File: exporters/otlp/common/build.gradle.kts (hunk 6 above)
Comment (jack-berg) [reply exporters/otlp/common/build.gradle.kts:12]: The `moduleName` is used to [set the location](https://github.com/open-telemetry/opentelemetry-java/blob/main/buildSrc/src/main/kotlin/otel.java-conventions.gradle.kts#L145-L158) of the `version.properties` file. It should reflect the base package of the artifact, so in this case it was wrong. If its not correct, then fetching a relative resource like `OtlpUserAgent.class.getResourceAsStream("version.properties")` will fail.

**Review by jkwatson — APPROVED** [review]

**Pull request merged**